File: app/vk_client.py
import logging
import os
import re
import time
from typing import Dict, List, Optional, Union

import requests
from dotenv import load_dotenv


logger = logging.getLogger("vk_feed.vk_client")


load_dotenv(override=False)


VK_TOKEN = os.getenv("VK_TOKEN", "")
VK_API_VERSION = os.getenv("VK_API_VERSION", "5.131")
VK_REQUEST_DELAY = float(os.getenv("VK_REQUEST_DELAY", "0.5"))


class VkApiError(Exception):
    pass


def _call_vk_method(method: str, params: Dict) -> Dict:
    url = f"https://api.vk.com/method/{method}"
    base_params = {
        "access_token": VK_TOKEN,
        "v": VK_API_VERSION,
    }
    merged_params = {**base_params, **params}

    for attempt in range(2):
        response = requests.get(url, params=merged_params, timeout=10)
        data = response.json()

        if "error" in data:
            error = data["error"]
            error_code = error.get("error_code")
            message = error.get("error_msg")
            logger.warning("VK API error %s: %s", error_code, message)

            if error_code == 6 and attempt == 0:
                time.sleep(2.0)
                continue

            raise VkApiError(f"VK API error {error_code}: {message}")

        return data

    raise VkApiError("VK API failed after retries")


def _extract_group_name_from_response(response: Dict, owner_id: int) -> Optional[str]:
    groups = response.get("response", {}).get("groups") or []
    for group in groups:
        if -int(group.get("id")) == owner_id:
            return group.get("name")
    return None


_VK_LINK_PATTERN = re.compile(r"\[(.+?)\|(.+?)\]")


def clean_vk_links(text: str) -> str:
    if not text:
        return ""
    return _VK_LINK_PATTERN.sub(r"\2", text)


def _map_attachments(raw_attachments: List[Dict]) -> List[Dict]:
    mapped: List[Dict] = []
    for att in raw_attachments or []:
        att_type = att.get("type")
        if not att_type:
            continue

        if att_type == "photo":
            photo = att.get("photo", {}) or {}
            sizes = photo.get("sizes") or []
            best_url = None
            if sizes:
                best = max(
                    sizes,
                    key=lambda s: (s.get("width") or 0) * (s.get("height") or 0),
                )
                best_url = best.get("url")
            mapped.append({"type": "photo", "url": best_url} if best_url else {"type": "photo"})
        elif att_type == "audio":
            audio = att.get("audio", {}) or {}
            artist = audio.get("artist") or ""
            title = audio.get("title") or ""
            full_title = f"{artist} - {title}".strip(" -")
            if full_title:
                mapped.append({"type": "audio", "title": full_title})
            else:
                mapped.append({"type": "audio"})
        elif att_type == "video":
            video = att.get("video", {}) or {}
            title = video.get("title") or ""
            if title:
                mapped.append({"type": "video", "title": title})
            else:
                mapped.append({"type": "video"})
        else:
            mapped.append({"type": att_type})

    return mapped


def fetch_posts_for_groups(groups: List[Union[str, Dict[str, str]]], count: int = 5) -> List[dict]:
    all_new_posts: List[dict] = []

    for group in groups:
        if isinstance(group, dict):
            group_id_str = str(group.get("id", "")).strip()
            group_name_from_config = str(group.get("name", "")).strip()
        else:
            group_id_str = str(group).strip()
            group_name_from_config = ""

        if not group_id_str:
            logger.warning("Empty group id value in config: %r", group)
            continue

        try:
            owner_id = int(group_id_str)
        except ValueError:
            logger.warning("Invalid group id value: %s", group_id_str)
            continue

        try:
            data = _call_vk_method(
                "wall.get",
                {
                    "owner_id": owner_id,
                    "count": count,
                    "extended": 1,
                },
            )
        except VkApiError as exc:
            logger.error("Failed to fetch posts for group %s: %s", group_id_str, exc)
            time.sleep(VK_REQUEST_DELAY)
            continue

        response = data.get("response", {})
        items = response.get("items", [])
        group_name_api = _extract_group_name_from_response(data, owner_id)
        group_name = group_name_from_config or group_name_api or str(group_id_str)

        for item in items:
            if "copy_history" in item:
                continue

            post_id = f'{item.get("owner_id")}_{item.get("id")}'
            attachments_raw = item.get("attachments", []) or []
            attachments = _map_attachments(attachments_raw)

            text_raw = item.get("text") or ""
            text_clean = clean_vk_links(text_raw)

            all_new_posts.append(
                {
                    "id": post_id,
                    "group": group_name,
                    "date": item.get("date"),
                    "text": text_clean,
                    "url": f'https://vk.com/wall{post_id}',
                    "attachments": attachments,
                }
            )

        time.sleep(VK_REQUEST_DELAY)

    return all_new_posts

File: src/lib/vk-client.js
import { appConfig } from './config.js'

const vkLinkPattern = /\[(.+?)\|(.+?)\]/g

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms))

export const cleanVkLinks = (text) => (text ? text.replace(vkLinkPattern, '$2') : '')

const callVkMethod = async (method, params) => {
  const url = new URL(`https://api.vk.com/method/${method}`)
  const merged = {
    access_token: appConfig.vkToken,
    v: appConfig.vkApiVersion,
    ...params,
  }
  Object.entries(merged).forEach(([key, value]) => url.searchParams.set(key, String(value)))

  for (let attempt = 0; attempt < 2; attempt += 1) {
    const response = await fetch(url, { cache: 'no-store' })
    const data = await response.json()
    if (!data.error) {
      return data
    }
    if (data.error.error_code === 6 && attempt === 0) {
      await sleep(2000)
      continue
    }
    throw new Error(`VK API error ${data.error.error_code}: ${data.error.error_msg}`)
  }
  throw new Error('VK API failed after retries')
}

const pickPhotoUrl = (photo) => {
  if (!photo || typeof photo !== 'object') {
    return ''
  }
  if (photo.orig_photo?.url) {
    return photo.orig_photo.url
  }
  const legacyKeys = ['photo_2560', 'photo_1280', 'photo_807', 'photo_604', 'photo_130', 'photo_75']
  for (const key of legacyKeys) {
    const candidate = photo[key]
    if (typeof candidate === 'string' && candidate.startsWith('http')) {
      return candidate
    }
  }
  const sizes = Array.isArray(photo.sizes) ? photo.sizes : []
  const best = sizes.reduce(
    (acc, item) => {
      const score = (item?.width || 0) * (item?.height || 0)
      return score > acc.score ? { score, url: item?.url || '' } : acc
    },
    { score: 0, url: '' },
  )
  return best.url || ''
}

const mapAttachments = (rawAttachments = []) => {
  return rawAttachments.map((att) => {
    const type = att?.type
    if (type === 'photo') {
      const url = pickPhotoUrl(att?.photo)
      return url ? { type: 'photo', url } : { type: 'photo' }
    }
    if (type === 'posted_photo') {
      const url = pickPhotoUrl(att?.posted_photo)
      return url ? { type: 'photo', url } : { type: 'photo' }
    }
    if (type === 'audio') {
      const artist = att?.audio?.artist || ''
      const title = att?.audio?.title || ''
      const full = `${artist} - ${title}`.trim().replace(/^- | -$/g, '')
      return full ? { type: 'audio', title: full } : { type: 'audio' }
    }
    if (type === 'video') {
      const title = att?.video?.title || ''
      return title ? { type: 'video', title } : { type: 'video' }
    }
    return { type }
  })
}

export const fetchPostsForGroups = async (groups, count = 5) => {
  const allPosts = []

  for (const group of groups) {
    const groupId = String(group?.id || group || '').trim()
    const groupNameFromConfig = String(group?.name || '').trim()
    if (!groupId) {
      continue
    }

    const ownerId = Number.parseInt(groupId, 10)
    if (!Number.isInteger(ownerId)) {
      continue
    }

    try {
      const data = await callVkMethod('wall.get', { owner_id: ownerId, count, extended: 1 })
      const response = data.response || {}
      const items = response.items || []
      const apiGroup = (response.groups || []).find((item) => Number(item.id) * -1 === ownerId)
      const groupName = groupNameFromConfig || apiGroup?.name || groupId

      items.forEach((item) => {
        if (item.copy_history) {
          return
        }
        const postId = `${item.owner_id}_${item.id}`
        allPosts.push({
          id: postId,
          group: groupName,
          date: item.date,
          text: cleanVkLinks(item.text || ''),
          url: `https://vk.com/wall${postId}`,
          attachments: mapAttachments(item.attachments || []),
        })
      })
    } catch {
      await sleep(appConfig.vkRequestDelayMs)
      continue
    }

    await sleep(appConfig.vkRequestDelayMs)
  }

  return allPosts
}
